contracts/guilds/GuildContract.cairo

```diff
@@ -0,0 +1,1351 @@
+end
+
+@storage_var
+func _is_whitelisted(account: felt) -> (res: felt):
+end
+
+@storage_var
+func _is_blacklisted(account: felt) -> (res: felt):
+end
+
+@storage_var
+func _permissions_len() -> (res: felt):
+end
+
+@storage_var
+func _permissions(index: felt) -> (res: Permission):
+end
+
+@storage_var
+func _guild_certificate() -> (res: felt):
+end
+
+@storage_var
+func _current_nonce() -> (res: felt):
+end
+
+#
+# Guards
+#
+
+func require_master{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    alloc_locals
+    let (caller) = get_caller_address()
+    let (master) = _guild_master.read()
+
+    with_attr error_message("Caller is not guild master"):
+        assert caller = master
+    end
+    return ()
+end
+
+func require_admin{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    alloc_locals
+    let (caller_address) = get_caller_address()
+    let (contract_address) = get_contract_address()
+    let (guild_certificate) = _guild_certificate.read()
+
+    let (certificate_id: Uint256) = IGuildCertificate.get_certificate_id(
+        contract_address=guild_certificate,
+        owner=caller_address,
+        guild=contract_address
+    )
+
+    let (_role) = IGuildCertificate.get_role(
+        contract_address=guild_certificate, 
+        certificate_id=certificate_id
+    )
+
+    with_attr error_message("Caller is not admin"):
+        assert _role = Role.ADMIN
+    end
+    return ()
+end
+
+func require_owner_or_member{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    alloc_locals
+    let (caller_address) = get_caller_address()
+    let (contract_address) = get_contract_address()
+    let (guild_certificate) = _guild_certificate.read()
+
+    let (certificate_id: Uint256) = IGuildCertificate.get_certificate_id(
+        contract_address=guild_certificate,
+        owner=caller_address,
+        guild=contract_address
+    )
+
+    let (check) = uint256_lt(Uint256(0,0),certificate_id)
+
+    with_attr error_mesage("Caller must have access"):
+        assert check = TRUE
+    end
+    return ()
+end
+
+func require_admin_or_owner{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }():
+    alloc_locals
+    let (caller_address) = get_caller_address()
+    let (contract_address) = get_contract_address()
+    let (guild_certificate) = _guild_certificate.read()
+
+    let (certificate_id: Uint256) = IGuildCertificate.get_certificate_id(
+        contract_address=guild_certificate,
+        owner=caller_address,
+        guild=contract_address
+    )
+
+    let (_role) = IGuildCertificate.get_role(
+        contract_address=guild_certificate, 
+        certificate_id=certificate_id
+    )
+
+    with_attr error_message("Caller is not admin or owner"):
+        assert_lt(1, _role)

```

Comment:
This check should be explicit. Not just `_role > 1`, but compare it to `OWNER` and `ADMIN`. What if the roles change in the future? We can't rely on the sequence.

Reply:
Great point, I will think of a better implementation of these guards

Reply:
In a different project I'm working on, we've built a bitwise access control list, which I think would fit the use case here quite well. We want to open source it eventually (maybe as soon as next week?), but I don't want to force you into it and if you want to come up with a solution without an external dependency, please, go ahead 😎

Reply:
Always open to suggestions. If you could link me the repo when it's out or fill me in with the details that would be awesome